src/iop/colorbalance.c

```diff
@@ -548,19 +1369,625 @@ static gboolean dt_iop_area_draw(GtkWidget *widget, cairo_t *cr, dt_iop_module_t
+    float rgb[3] = { 0.0f };
+    dt_Lab_to_XYZ((const float *)self->picked_color, XYZ);
+    dt_XYZ_to_prophotorgb((const float *)XYZ, rgb);
+    
+    const float lift[3] = { ( p->lift[CHANNEL_RED] + p->lift[CHANNEL_FACTOR] - 2.0f ),
+                              ( p->lift[CHANNEL_GREEN] + p->lift[CHANNEL_FACTOR] - 2.0f ),
+                              ( p->lift[CHANNEL_BLUE] + p->lift[CHANNEL_FACTOR] - 2.0f ) },
+                gamma[3] = { p->gamma[CHANNEL_RED] * p->gamma[CHANNEL_FACTOR],
+                             p->gamma[CHANNEL_GREEN] * p->gamma[CHANNEL_FACTOR],
+                             p->gamma[CHANNEL_BLUE] * p->gamma[CHANNEL_FACTOR] },
+                gamma_inv[3] = { (gamma[0] != 0.0f) ? 1.0f / gamma[0] : 1000000.0f,
+                                 (gamma[1] != 0.0f) ? 1.0f / gamma[1] : 1000000.0f,
+                                 (gamma[2] != 0.0f) ? 1.0f / gamma[2] : 1000000.0f },
+                gain[3] = { p->gain[CHANNEL_RED] * p->gain[CHANNEL_FACTOR],
+                            p->gain[CHANNEL_GREEN] * p->gain[CHANNEL_FACTOR],
+                            p->gain[CHANNEL_BLUE] * p->gain[CHANNEL_FACTOR] };
+                
+    for(int c = 0; c < 3; c++)
+    {
+      rgb[c] = CDL(rgb[c], gain[c], lift[c], gamma_inv[c]);
+      rgb[c] = CLAMP(rgb[c], 0.0f, 1.0f);
+    }
+  
+    dt_prophotorgb_to_XYZ((const float *)rgb, XYZ);
+    
+    p->grey = XYZ[1] * 100.0f;
+
+    dt_bauhaus_slider_set_soft(g->grey, p->grey);
+    
+    self->request_color_pick = DT_REQUEST_COLORPICK_OFF;
+    
+    dt_dev_add_history_item(darktable.develop, self, TRUE);
+  }
+}
+
+static void lift_neutralize_callback(GtkWidget *button, gpointer user_data)
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  if(self->dt->gui->reset) return;
+  
+  dt_iop_request_focus(self);
+  dt_control_queue_redraw();
+
+  if(self->request_color_pick == DT_REQUEST_COLORPICK_MODULE)
+  {
+    dt_dev_reprocess_all(self->dev);
+  }
+
+  if(self->request_color_pick == DT_REQUEST_COLORPICK_OFF)
+    self->request_color_pick = DT_REQUEST_COLORPICK_MODULE;
+  else
+  {
+    dt_iop_colorbalance_params_t *p = (dt_iop_colorbalance_params_t *)self->params;
+    dt_iop_colorbalance_gui_data_t *g = (dt_iop_colorbalance_gui_data_t *)self->gui_data;
+    
+    if(self->request_color_pick != DT_REQUEST_COLORPICK_MODULE || self->picked_color_max[0] < 0.0f) 
+    {
+      dt_control_log(_("wait for the preview to be updated."));
+      return;
+    }
+    
+    const float gain[3] = { p->gain[CHANNEL_RED],
+                            p->gain[CHANNEL_GREEN],
+                            p->gain[CHANNEL_BLUE] };
+    
+    float XYZ[3] = { 0.0f };
+    dt_Lab_to_XYZ((const float *)self->picked_color, XYZ);
+    float RGB[3] = { 0.0f };
+    dt_XYZ_to_prophotorgb((const float *)XYZ, RGB);
+    
+    // Save the patch color for the optimization
+    dt_iop_colorbalance_data_t *d = (dt_iop_colorbalance_data_t *)self->data;
+    *d->color_patches_lift = *RGB;
```

Comment:
4 floats
`float color_patches_lift[CHANNEL_SIZE];
`
3 floats:

`float RGB[3] = { 0.0f };
`

So not sure the "*left = *right" is correct. There may be some other wrong copy like this.

Reply:
` float color_patches_lift[CHANNEL_SIZE];`

if enum begins @ 0, `CHANNEL_SIZE` == 3, isn't it ?

Reply:
Of course not :) C is stupid. That's a quite old language at this stage and with lot of pitfall for us to kill ourselves:)